core/base/dispatch_helper.hpp

```diff
@@ -120,18 +126,49 @@ ReturnType run_impl(T obj, Func&& f, Args&&... args)
  */
 template <typename K, typename... Types, typename T, typename Func,
           typename... Args>
-auto run(T obj, Func&& f, Args&&... args)
+auto run(T* obj, Func&& f, Args&&... args)
 {
 #if __cplusplus < 201703L
-    using ReturnType = std::result_of_t<Func(K, Args...)>;
+    using ReturnType =
+        std::result_of_t<Func(detail::with_same_constness_t<K, T>*, Args...)>;
 #else
-    using ReturnType = std::invoke_result_t<Func, K, Args...>;
+    using ReturnType =
+        std::invoke_result_t<Func, detail::with_same_constness_t<K, T>*,
+                             Args...>;
 #endif
     return detail::run_impl<ReturnType, K, Types...>(
         obj, std::forward<Func>(f), std::forward<Args>(args)...);
 }
 
 
+/**
+ * run uses template to go through the list and select the valid
+ * template and run it.
+ *
+ * @tparam Base  the Base class with one template
+ * @tparam K  the current template type of B. pointer of const Base<K> is tried
+ *            in the conversion.
+ * @tparam ...Types  the other types will be tried in the conversion if K fails
+ * @tparam T  the type of input object waiting converted
+ * @tparam Func  the function will run if the object can be converted to pointer
+ *               of const Base<K>
+ * @tparam ...Args  the additional arguments for the Func
+ *
+ * @param obj  the input object waiting converted
+ * @param f  the function will run if obj can be converted successfully
+ * @param args  the additional arguments for the function
+ *
+ * @return  the result of f invoked with obj cast to the first matching type
+ */
+template <template <class> class K, typename... Types, typename T,
```

Comment:
```suggestion
template <template <class> class Base, typename... Types, typename T,
```
also the documentation needs to be adjusted. I have changed some documentations here, you may need to rebase and keep them similar.
With that, I think the following function can be deleted, right?

Reply:
Which following function do you mean? This at l.221 or at l.252?